### PR TITLE
make impersonateSAR with empty token illegal

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/4529.

This does not fix the cert-auth hole.  If your base client is authenticated using certificate authentication and you've specified a token, the certificate is ignored.  I see a couple possibilities:
 1. leave it as is (this is better than no change) and take an issue to fix outside of this change
 1. remove this API and make the caller say what they mean (I like this)
 1. inspect the roundtrippers and make sure that we don't have a client cert wrapper.  If we do panic.  This is a pain in usage, but relatively safe for now.

@liggitt

### CODE REVIEW

File: pkg/client/subjectaccessreview.go [hunk 1]
```diff
@@ -76,9 +76,10 @@ func (c *subjectAccessReviews) Create(sar *authorizationapi.SubjectAccessReview)
 	return
 }
 
-func overrideAuth(token string, req *client.Request) *client.Request {
-	if len(token) > 0 {
-		req.SetHeader("Authorization", fmt.Sprintf("Bearer %s", token))
+// overrideAuth specifies the token to authenticate the request with.  This only works if certificate auth isn't also being used.
+func overrideAuth(token *string, req *client.Request) *client.Request {
+	if token != nil {
```
Comment (liggitt) [inline pkg/client/subjectaccessreview.go]: return an error if token is non-nil but empty?

Comment (liggitt) [inline pkg/client/subjectaccessreview.go]: I'd like to avoid that hitting the server...

**Comment by deads2k** [comment]
Comment addressed.  Of course, that only prevents our client code from doing it, it doesn't actually let the server be lazy.

**Comment by liggitt** [comment]
lgtm, [merge]

**Pull request merged**